test/StateDiffChecker.t.sol

```diff
@@ -0,0 +1,190 @@
+    {
+        vm.startStateDiffRecording();
+        target.set();
+        VmSafe.AccountAccess[] memory accountAccesses = vm.stopAndReturnStateDiff();
+
+        string memory _path = "test/testDiff.json";
+        string memory json = vm.readFile(_path);
+        expectedDiff_ = Checker.parseDiffSpecs(json);
+
+        actualDiff_ = Checker.extractDiffSpecFromAcccountAccesses(accountAccesses);
+    }
+
+    /// @dev Test that the state diff generated by running Target.set() matches the expected state diff
+    ///      as defined in testDiff.json
+    function test_checkStateDiff_succeeds() public {
+        (Checker.StateDiffSpec memory expectedDiff, Checker.StateDiffSpec memory actualDiff) = _executeAndGetDiffs();
+        Checker.checkStateDiff(expectedDiff, actualDiff);
+    }
+
+    /// @dev Test that the correct error is thrown when the chain ID does not match.
+    function test_checkStateDiff_chainIdMismatch_reverts() public {
+        (Checker.StateDiffSpec memory expectedDiff, Checker.StateDiffSpec memory actualDiff) = _executeAndGetDiffs();
+        actualDiff.chainId = 31338;
+        vm.expectRevert(abi.encodeWithSelector(StateDiffMismatch.selector, "chainId", 31337, 31338));
+        Checker.checkStateDiff(expectedDiff, actualDiff);
+    }
+
+    /// @dev Test that the correct error is thrown when the number of storage modifications does not match.
+    function test_checkStateDiff_lengthMismatch_reverts() public {
+        (Checker.StateDiffSpec memory expectedDiff,) = _executeAndGetDiffs();
+        Checker.StateDiffSpec memory shorterDiff = Checker.StateDiffSpec({
+            chainId: 31337,
+            storageSpecs: new Checker.StorageDiffSpec[](expectedDiff.storageSpecs.length - 1)
+        });
+        shorterDiff.storageSpecs[0] = expectedDiff.storageSpecs[0];
+        vm.expectRevert(
+            abi.encodeWithSelector(
+                StateDiffMismatch.selector,
+                "storageSpecs.length",
+                expectedDiff.storageSpecs.length,
+                shorterDiff.storageSpecs.length
+            )
+        );
+        Checker.checkStateDiff(expectedDiff, shorterDiff);
+    }
+
+    /// @dev A utility function to copy a StateDiffSpec struct so that we can modify it without affecting the original.
+    function copyDiff(Checker.StateDiffSpec memory diff) internal pure returns (Checker.StateDiffSpec memory) {
+        Checker.StateDiffSpec memory copy;
+        copy.chainId = diff.chainId;
+        copy.storageSpecs = new Checker.StorageDiffSpec[](diff.storageSpecs.length);
+        for (uint256 i = 0; i < diff.storageSpecs.length; i++) {
+            copy.storageSpecs[i].account = diff.storageSpecs[i].account;
+            copy.storageSpecs[i].slot = diff.storageSpecs[i].slot;
+            copy.storageSpecs[i].newValue = diff.storageSpecs[i].newValue;
+            copy.storageSpecs[i].previousValue = diff.storageSpecs[i].previousValue;
+        }
+        return copy;
```

Comment:
A downside here is that if we add fields to the state diff we have to manually remember to update this function. Any ideas on a good alternative here so this is easier to maintain? I think copying with assembly like below would copy by reference
```solidity
Checker.StateDiffSpec memory copy;
assembly {
  copy := diff
}
return copy;
```

The full array should be sequential in memory so maybe we can extract the length out and loop over those memory chunks (or use the identity precompile) to copy that to a new memory location?

Reply:
Idea from @maurelian that we'll go with: just verify that the hash of `copy` matches the hash of `diff` and revert if not

Reply:
I implemented this.